File: test/VbNet.LanguageServer.Tests/Integration/DiagnosticsIntegrationTests.cs
using Microsoft.Extensions.Logging.Abstractions;
using VbNet.LanguageServer.Protocol;
using VbNet.LanguageServer.Services;
using VbNet.LanguageServer.Workspace;
using Xunit;

namespace VbNet.LanguageServer.Tests.Integration;

/// <summary>
/// Integration tests for DiagnosticsService with real VB.NET projects.
/// These tests verify that diagnostics are correctly computed from Roslyn.
/// </summary>
[Collection("MSBuild")]
public class DiagnosticsIntegrationTests : IAsyncLifetime
{
    private readonly WorkspaceManager _workspaceManager;
    private readonly DocumentManager _documentManager;
    private readonly DiagnosticsService _diagnosticsService;
    private readonly List<PublishDiagnosticsParams> _publishedDiagnostics = new();

    private static readonly string TestProjectsRoot = GetTestProjectsRoot();

    public DiagnosticsIntegrationTests()
    {
        _workspaceManager = new WorkspaceManager(NullLogger<WorkspaceManager>.Instance);
        _documentManager = new DocumentManager(_workspaceManager, NullLogger<DocumentManager>.Instance);
        _diagnosticsService = new DiagnosticsService(
            _workspaceManager,
            _documentManager,
            PublishDiagnosticsAsync,
            NullLogger<DiagnosticsService>.Instance);
    }

    private static string GetTestProjectsRoot()
    {
        var assemblyLocation = typeof(DiagnosticsIntegrationTests).Assembly.Location;
        var assemblyDir = Path.GetDirectoryName(assemblyLocation)!;
        var testProjectsPath = Path.GetFullPath(Path.Combine(assemblyDir, "..", "..", "..", "..", "TestProjects"));
        return testProjectsPath;
    }

    private Task PublishDiagnosticsAsync(string method, PublishDiagnosticsParams @params, CancellationToken ct)
    {
        _publishedDiagnostics.Add(@params);
        return Task.CompletedTask;
    }

    public Task InitializeAsync()
    {
        _workspaceManager.Initialize();
        return Task.CompletedTask;
    }

    public async Task DisposeAsync()
    {
        _diagnosticsService.Dispose();
        await _workspaceManager.DisposeAsync();
    }

    [Fact]
    public async Task GetDiagnosticsAsync_ValidCode_ReturnsNoDiagnostics()
    {
        var projectPath = Path.Combine(TestProjectsRoot, "SmallProject", "SmallProject.vbproj");
        var helperPath = Path.Combine(TestProjectsRoot, "SmallProject", "Helper.vb");

        if (!File.Exists(projectPath))
        {
            return;
        }

        await _workspaceManager.LoadProjectAsync(projectPath);

        var helperUri = new Uri(helperPath).ToString();

        // Open the document
        var text = await File.ReadAllTextAsync(helperPath);
        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = helperUri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        // Get diagnostics directly (bypass debounce)
        var diagnostics = await _diagnosticsService.GetDiagnosticsAsync(helperUri);

        // Helper.vb should have no errors
        var errors = diagnostics.Where(d => d.Severity == DiagnosticSeverity.Error).ToList();
        Assert.Empty(errors);
    }

    [Fact]
    public async Task GetDiagnosticsAsync_ErrorCode_ReturnsDiagnostics()
    {
        var projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj");
        var errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb");

        if (!File.Exists(projectPath))
        {
            return;
        }

        await _workspaceManager.LoadProjectAsync(projectPath);

        var errorClassUri = new Uri(errorClassPath).ToString();

        // Open the document
        var text = await File.ReadAllTextAsync(errorClassPath);
        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = errorClassUri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        // Get diagnostics
        var diagnostics = await _diagnosticsService.GetDiagnosticsAsync(errorClassUri);

        // ErrorClass.vb should have multiple errors
        var errors = diagnostics.Where(d => d.Severity == DiagnosticSeverity.Error).ToList();
        Assert.NotEmpty(errors);

        // Check for specific error codes
        var errorCodes = errors.Select(e => e.Code).ToList();

        // BC30512 - Option Strict On disallows implicit conversions (string to integer)
        Assert.Contains("BC30512", errorCodes);

        // BC30451 - Undefined variable
        Assert.Contains("BC30451", errorCodes);

        // BC30002 - Type undefined
        Assert.Contains("BC30002", errorCodes);
    }

    [Fact]
    public async Task GetDiagnosticsAsync_DiagnosticsHaveCorrectRanges()
    {
        var projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj");
        var errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb");

        if (!File.Exists(projectPath))
        {
            return;
        }

        await _workspaceManager.LoadProjectAsync(projectPath);

        var errorClassUri = new Uri(errorClassPath).ToString();
        var text = await File.ReadAllTextAsync(errorClassPath);

        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = errorClassUri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        var diagnostics = await _diagnosticsService.GetDiagnosticsAsync(errorClassUri);

        // All diagnostics should have valid ranges
        foreach (var diagnostic in diagnostics)
        {
            Assert.NotNull(diagnostic.Range);
            Assert.True(diagnostic.Range.Start.Line >= 0);
            Assert.True(diagnostic.Range.Start.Character >= 0);
            Assert.True(diagnostic.Range.End.Line >= diagnostic.Range.Start.Line);
        }
    }

    [Fact]
    public async Task GetDiagnosticsAsync_DiagnosticsHaveSource()
    {
        var projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj");
        var errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb");

        if (!File.Exists(projectPath))
        {
            return;
        }

        await _workspaceManager.LoadProjectAsync(projectPath);

        var errorClassUri = new Uri(errorClassPath).ToString();
        var text = await File.ReadAllTextAsync(errorClassPath);

        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = errorClassUri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        var diagnostics = await _diagnosticsService.GetDiagnosticsAsync(errorClassUri);

        // All diagnostics should have source set to "vbnet"
        Assert.All(diagnostics, d => Assert.Equal("vbnet", d.Source));
    }

    [Fact]
    public async Task ComputeAndPublishDiagnosticsAsync_PublishesDiagnostics()
    {
        var projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj");
        var errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb");

        if (!File.Exists(projectPath))
        {
            return;
        }

        await _workspaceManager.LoadProjectAsync(projectPath);

        var errorClassUri = new Uri(errorClassPath).ToString();
        var text = await File.ReadAllTextAsync(errorClassPath);

        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = errorClassUri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        _publishedDiagnostics.Clear();

        // Compute and publish diagnostics
        await _diagnosticsService.ComputeAndPublishDiagnosticsAsync(errorClassUri);

        // Should have published diagnostics
        Assert.Single(_publishedDiagnostics);
        Assert.Equal(errorClassUri, _publishedDiagnostics[0].Uri);
        Assert.NotEmpty(_publishedDiagnostics[0].Diagnostics);
    }

    [Fact]
    public async Task ClearDiagnosticsAsync_PublishesEmptyDiagnostics()
    {
        var projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj");
        var errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb");

        if (!File.Exists(projectPath))
        {
            return;
        }

        await _workspaceManager.LoadProjectAsync(projectPath);

        var errorClassUri = new Uri(errorClassPath).ToString();
        var text = await File.ReadAllTextAsync(errorClassPath);

        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = errorClassUri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        // First publish some diagnostics
        await _diagnosticsService.ComputeAndPublishDiagnosticsAsync(errorClassUri);

        _publishedDiagnostics.Clear();

        // Clear diagnostics
        await _diagnosticsService.ClearDiagnosticsAsync(errorClassUri);

        // Should have published empty diagnostics
        Assert.Single(_publishedDiagnostics);
        Assert.Equal(errorClassUri, _publishedDiagnostics[0].Uri);
        Assert.Empty(_publishedDiagnostics[0].Diagnostics);
    }

    [Fact]
    public async Task DocumentReassociation_TriggersDiagnostics()
    {
        var projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj");
        var errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb");

        if (!File.Exists(projectPath))
        {
            return;
        }

        var errorClassUri = new Uri(errorClassPath).ToString();
        var text = await File.ReadAllTextAsync(errorClassPath);

        // Open document BEFORE loading workspace (simulates real client behavior)
        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = errorClassUri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        // Document should not have DocumentId yet
        var doc = _documentManager.GetOpenDocument(errorClassUri);
        Assert.NotNull(doc);
        Assert.Null(doc.DocumentId);

        _publishedDiagnostics.Clear();

        // Now load the workspace - this should trigger reassociation and diagnostics
        await _workspaceManager.LoadProjectAsync(projectPath);

        // Document should now have DocumentId
        doc = _documentManager.GetOpenDocument(errorClassUri);
        Assert.NotNull(doc);
        Assert.NotNull(doc.DocumentId);

        // Wait a bit for debounced diagnostics (300ms default + buffer)
        await Task.Delay(500);

        // Should have received diagnostics after reassociation
        // Note: This may or may not have been published depending on debounce timing
        // The key assertion is that the document was reassociated
    }

    [Fact]
    public async Task MinimumSeverity_FiltersCorrectly()
    {
        var projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj");
        var errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb");

        if (!File.Exists(projectPath))
        {
            return;
        }

        await _workspaceManager.LoadProjectAsync(projectPath);

        var errorClassUri = new Uri(errorClassPath).ToString();
        var text = await File.ReadAllTextAsync(errorClassPath);

        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = errorClassUri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        // Set minimum severity to Error only
        _diagnosticsService.MinimumSeverity = DiagnosticSeverity.Error;

        var diagnostics = await _diagnosticsService.GetDiagnosticsAsync(errorClassUri);

        // Should only have errors, no warnings or hints
        Assert.All(diagnostics, d => Assert.Equal(DiagnosticSeverity.Error, d.Severity));
    }
}

File: test/VbNet.LanguageServer.Tests.Vb/Integration/DiagnosticsIntegrationTests.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Threading
Imports System.Threading.Tasks
Imports Microsoft.Extensions.Logging.Abstractions
Imports VbNet.LanguageServer.Protocol
Imports VbNet.LanguageServer.Services
Imports VbNet.LanguageServer.Workspace
Imports Xunit

Namespace VbNet.LanguageServer.Tests.Integration

    ''' <summary>
    ''' Integration tests for DiagnosticsService with real VB.NET projects.
    ''' These tests verify that diagnostics are correctly computed from Roslyn.
    ''' </summary>
    <Collection("MSBuild")>
    Public Class DiagnosticsIntegrationTests
        Implements IAsyncLifetime

        Private ReadOnly _workspaceManager As WorkspaceManager
        Private ReadOnly _documentManager As DocumentManager
        Private ReadOnly _diagnosticsService As DiagnosticsService
        Private ReadOnly _publishedDiagnostics As New List(Of PublishDiagnosticsParams)()

        Private Shared ReadOnly TestProjectsRoot As String = GetTestProjectsRoot()

        Public Sub New()
            _workspaceManager = New WorkspaceManager(NullLogger(Of WorkspaceManager).Instance)
            _documentManager = New DocumentManager(_workspaceManager, NullLogger(Of DocumentManager).Instance)
            _diagnosticsService = New DiagnosticsService(
                _workspaceManager,
                _documentManager,
                AddressOf PublishDiagnosticsAsync,
                NullLogger(Of DiagnosticsService).Instance)
        End Sub

        Private Shared Function GetTestProjectsRoot() As String
            Dim assemblyLocation = GetType(DiagnosticsIntegrationTests).Assembly.Location
            Dim assemblyDir = Path.GetDirectoryName(assemblyLocation)
            Dim testProjectsPath = Path.GetFullPath(Path.Combine(assemblyDir, "..", "..", "..", "..", "TestProjects"))
            Return testProjectsPath
        End Function

        Private Function PublishDiagnosticsAsync(method As String, paramsValue As PublishDiagnosticsParams, ct As CancellationToken) As Task
            _publishedDiagnostics.Add(paramsValue)
            Return Task.CompletedTask
        End Function

        Public Function InitializeAsync() As Task Implements IAsyncLifetime.InitializeAsync
            _workspaceManager.Initialize()
            Return Task.CompletedTask
        End Function

        Public Async Function DisposeAsync() As Task Implements IAsyncLifetime.DisposeAsync
            _diagnosticsService.Dispose()
            Await _workspaceManager.DisposeAsync()
        End Function

        <Fact>
        Public Async Function GetDiagnosticsAsync_ValidCode_ReturnsNoDiagnostics() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "SmallProject", "SmallProject.vbproj")
            Dim helperPath = Path.Combine(TestProjectsRoot, "SmallProject", "Helper.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim helperUri = New Uri(helperPath).ToString()
            Dim text = Await File.ReadAllTextAsync(helperPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = helperUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim diagnostics = Await _diagnosticsService.GetDiagnosticsAsync(helperUri)

            Dim errors = diagnostics.Where(Function(d) d.Severity = DiagnosticSeverity.Error).ToList()
            Assert.Empty(errors)
        End Function

        <Fact>
        Public Async Function GetDiagnosticsAsync_ErrorCode_ReturnsDiagnostics() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj")
            Dim errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim errorClassUri = New Uri(errorClassPath).ToString()
            Dim text = Await File.ReadAllTextAsync(errorClassPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = errorClassUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim diagnostics = Await _diagnosticsService.GetDiagnosticsAsync(errorClassUri)

            Dim errors = diagnostics.Where(Function(d) d.Severity = DiagnosticSeverity.Error).ToList()
            Assert.NotEmpty(errors)

            Dim errorCodes = errors.Select(Function(e) e.Code).ToList()

            Assert.Contains("BC30512", errorCodes)
            Assert.Contains("BC30451", errorCodes)
            Assert.Contains("BC30002", errorCodes)
        End Function

        <Fact>
        Public Async Function GetDiagnosticsAsync_DiagnosticsHaveCorrectRanges() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj")
            Dim errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim errorClassUri = New Uri(errorClassPath).ToString()
            Dim text = Await File.ReadAllTextAsync(errorClassPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = errorClassUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim diagnostics = Await _diagnosticsService.GetDiagnosticsAsync(errorClassUri)

            For Each diagnostic In diagnostics
                Assert.NotNull(diagnostic.Range)
                Assert.True(diagnostic.Range.Start.Line >= 0)
                Assert.True(diagnostic.Range.Start.Character >= 0)
                Assert.True(diagnostic.Range.End.Line >= diagnostic.Range.Start.Line)
            Next
        End Function

        <Fact>
        Public Async Function GetDiagnosticsAsync_DiagnosticsHaveSource() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj")
            Dim errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim errorClassUri = New Uri(errorClassPath).ToString()
            Dim text = Await File.ReadAllTextAsync(errorClassPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = errorClassUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim diagnostics = Await _diagnosticsService.GetDiagnosticsAsync(errorClassUri)

            Assert.All(diagnostics, Sub(d) Assert.Equal("vbnet", d.Source))
        End Function

        <Fact>
        Public Async Function ComputeAndPublishDiagnosticsAsync_PublishesDiagnostics() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj")
            Dim errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim errorClassUri = New Uri(errorClassPath).ToString()
            Dim text = Await File.ReadAllTextAsync(errorClassPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = errorClassUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            _publishedDiagnostics.Clear()

            Await _diagnosticsService.ComputeAndPublishDiagnosticsAsync(errorClassUri)

            Assert.Single(_publishedDiagnostics)
            Assert.Equal(errorClassUri, _publishedDiagnostics(0).Uri)
            Assert.NotEmpty(_publishedDiagnostics(0).Diagnostics)
        End Function

        <Fact>
        Public Async Function ClearDiagnosticsAsync_PublishesEmptyDiagnostics() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj")
            Dim errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim errorClassUri = New Uri(errorClassPath).ToString()
            Dim text = Await File.ReadAllTextAsync(errorClassPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = errorClassUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Await _diagnosticsService.ComputeAndPublishDiagnosticsAsync(errorClassUri)

            _publishedDiagnostics.Clear()

            Await _diagnosticsService.ClearDiagnosticsAsync(errorClassUri)

            Assert.Single(_publishedDiagnostics)
            Assert.Equal(errorClassUri, _publishedDiagnostics(0).Uri)
            Assert.Empty(_publishedDiagnostics(0).Diagnostics)
        End Function

        <Fact>
        Public Async Function DocumentReassociation_TriggersDiagnostics() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj")
            Dim errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Dim errorClassUri = New Uri(errorClassPath).ToString()
            Dim text = Await File.ReadAllTextAsync(errorClassPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = errorClassUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim doc = _documentManager.GetOpenDocument(errorClassUri)
            Assert.NotNull(doc)
            Assert.Null(doc.DocumentId)

            _publishedDiagnostics.Clear()

            Await _workspaceManager.LoadProjectAsync(projectPath)

            doc = _documentManager.GetOpenDocument(errorClassUri)
            Assert.NotNull(doc)
            Assert.NotNull(doc.DocumentId)

            Await Task.Delay(500)
        End Function

        <Fact>
        Public Async Function MinimumSeverity_FiltersCorrectly() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj")
            Dim errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim errorClassUri = New Uri(errorClassPath).ToString()
            Dim text = Await File.ReadAllTextAsync(errorClassPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = errorClassUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            _diagnosticsService.MinimumSeverity = DiagnosticSeverity.Error

            Dim diagnostics = Await _diagnosticsService.GetDiagnosticsAsync(errorClassUri)

            Assert.All(diagnostics, Sub(d) Assert.Equal(DiagnosticSeverity.Error, d.Severity))
        End Function

        <Fact>
        Public Async Function GetDocumentDiagnosticsReportAsync_ReturnsDiagnostics() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj")
            Dim errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim errorClassUri = New Uri(errorClassPath).ToString()
            Dim text = Await File.ReadAllTextAsync(errorClassPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = errorClassUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim report = Await _diagnosticsService.GetDocumentDiagnosticsReportAsync(New TextDocumentDiagnosticParams With {
                .TextDocument = New TextDocumentIdentifier With {.Uri = errorClassUri}
            })

            Assert.NotNull(report)
            Assert.Equal("full", report.Kind)
            Assert.NotEmpty(report.Items)
        End Function

        <Fact>
        Public Async Function GetWorkspaceDiagnosticsReportAsync_ReturnsWorkspaceItems() As Task
            Dim projectPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorProject.vbproj")
            Dim errorClassPath = Path.Combine(TestProjectsRoot, "ErrorProject", "ErrorClass.vb")

            If Not File.Exists(projectPath) Then
                Return
            End If

            Await _workspaceManager.LoadProjectAsync(projectPath)

            Dim errorClassUri = New Uri(errorClassPath).ToString()
            Dim text = Await File.ReadAllTextAsync(errorClassPath)

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = errorClassUri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim report = Await _diagnosticsService.GetWorkspaceDiagnosticsReportAsync(New WorkspaceDiagnosticParams())

            Assert.NotNull(report)
            Assert.NotEmpty(report.Items)
            Assert.Contains(report.Items, Function(item) item.Uri = errorClassUri AndAlso item.Items.Length > 0)
        End Function
    End Class

End Namespace
